crates/bevy_ecs/src/query/filter.rs

```diff
@@ -543,36 +562,36 @@ macro_rules! impl_flag_filter {
                         self.entity_table_rows = archetype.entity_table_rows().as_ptr();
                         // SAFE: archetype tables always exist
                         let table = tables.get_unchecked(archetype.table_id());
-                        self.table_flags = table
+                        self.table_counters = table
                             .get_column(state.component_id).unwrap()
-                            .get_flags_mut_ptr();
+                            .get_counters_mut_ptr();
                     }
                     StorageType::SparseSet => self.entities = archetype.entities().as_ptr(),
                 }
             }
 
             unsafe fn table_fetch(&mut self, table_row: usize) -> bool {
-                false $(|| (*self.table_flags.add(table_row)).contains($flags))+
+                false $(|| $counters(&*self.table_counters.add(table_row), self.system_counter, self.global_system_counter))+
             }
 
             unsafe fn archetype_fetch(&mut self, archetype_index: usize) -> bool {
                 match self.storage_type {
                     StorageType::Table => {
                         let table_row = *self.entity_table_rows.add(archetype_index);
-                        false $(|| (*self.table_flags.add(table_row)).contains($flags))+
+                        false $(|| $counters(&*self.table_counters.add(table_row), self.system_counter, self.global_system_counter))+
                     }
                     StorageType::SparseSet => {
                         let entity = *self.entities.add(archetype_index);
-                        let flags = (*(*self.sparse_set).get_flags(entity).unwrap());
-                        false $(|| flags.contains($flags))+
+                        let counters = (*(*self.sparse_set).get_counters(entity).unwrap());
+                        false $(|| $counters(&counters, self.system_counter, self.global_system_counter))+
                     }
                 }
             }
         }
     };
 }
 
-impl_flag_filter!(
+impl_counter_filter!(
     /// Filter that retrieves components of type `T` that have been added since the start of the frame
     ///
     /// This filter is useful as a performance optimization as it means that the query contains fewer items
```

Comment:
This isn't the only reason why the filter is useful. `Added` in particular is useful for one-time post-processing.